### PR TITLE
GEOSClipByRect: preserve Z coordinate

### PR DESCRIPTION
**Pull request by vadimcn**
For [multi]linestrings/linerings, Z coordinate is interpolated from that of the nearest existing vertices.

For [multi]polygons, Z coordinate of points created on the boundary is interpolated as above.  For interior points, however, it is populated using the ElevationModel.

### CODE REVIEW

File: src/operation/intersection/RectangleIntersection.cpp [hunk 1]
```diff
@@ -682,7 +691,12 @@ std::unique_ptr<geom::Geometry>
 RectangleIntersection::clip(const geom::Geometry& g, const Rectangle& rect)
 {
     RectangleIntersection ri(g, rect);
-    return ri.clip();
+    std::unique_ptr<geom::Geometry> result = ri.clip();
+
+    std::unique_ptr<ElevationModel> elevModel = ElevationModel::create(g);
```
Comment (dbaston) [inline src/operation/intersection/RectangleIntersection.cpp:696]: This can be conditioned on the input actually having Z values.

Comment (vadimcn) [reply src/operation/intersection/RectangleIntersection.cpp]: done

**Comment by vadimcn** [comment]
@dbaston does this look okay to merge now?

**Comment by dbaston** [comment]
Do you think it makes sense to modify the tests to remove the expected value and just check that the result matches `GEOSIntersection` ?

**Comment by vadimcn** [comment]
I don't know enough about the GEOS project to have a strong opinion on this.
Do you postulate that the result of GEOSClipByRect should always be the same as GEOSIntersection with a box polygon?  Or should it be the same after normalization, as you have above?  Then maybe yes.

However, currently, several existing tests fail this comparison, see below.  I'd say that fixing those is out of scope of this PR.

EXP: POINT EMPTY
OBT: GEOMETRYCOLLECTION EMPTY
[1=F]

EXP: POINT (15 10)
OBT: GEOMETRYCOLLECTION EMPTY
[3=F]

EXP: LINESTRING EMPTY
OBT: GEOMETRYCOLLECTION EMPTY
[4=F]

EXP: MULTILINESTRING ((10 10, 15 10), (10 10, 10 15))
OBT: GEOMETRYCOLLECTION EMPTY
[6=F]

EXP: MULTILINESTRING ((20 10, 20 20), (10 20, 20 20), (10 10, 20 10), (10 10, 10 20))
OBT: GEOMETRYCOLLECTION EMPTY
[10=F]

EXP: MULTILINESTRING ((20 10, 20 20), (10 20, 20 20), (10 10, 20 10), (10 10, 10 20))
OBT: GEOMETRYCOLLECTION EMPTY
[11=F]

**Comment by dbaston** [comment]
> However, currently, several existing tests fail this comparison, see below.

Odd, they are passing for me after normalization. Either way, it can be beyond the scope of this PR.

**Comment by vadimcn** [comment]
> Odd, they are passing for me after normalization. Either way, it can be beyond the scope of this PR.

I thought the plan was to update *all* tests, including those still using `checkClipByRect`. If you only changed `checkClipByRectIdentical`, I guess it'd work.

**Comment by dbaston** [comment]
Looking more closely at the examples you pointed out, I can see that the semantics are different from `GEOSIntersection`. I guess the tests should be left as-is.

**Pull request merged**